src/plugins/mer/meremulatordevice.cpp

```diff
@@ -453,6 +454,56 @@ MerConnection *MerEmulatorDevice::connection() const
     return m_connection.data();
 }
 
+void MerEmulatorDevice::setPortForwardingRule(const QString& ruleName, quint16 hostPort,
+                                              quint16 emulatorVmPort) const {
+    MerVirtualBoxManager::updatePortForwardingRule(m_connection->virtualMachine(), ruleName,
+                                                   hostPort, emulatorVmPort);
+}
+
+/*!
+ * \brief Checks whether the port forwarding rule with the given host machine port is set on the
+ * emulator virtual machine or not.
+ * \param hostPort Port of the host machine to check.
+ * \return True if the port forwarding rule with the given host port is set on the emulator,
+ *         false — otherwise.
+ */
+bool MerEmulatorDevice::isPortForwardingSet(quint16 hostPort) const {
```

Comment:
Is synchronous operation really acceptable for your use case?

Reply:
The `MerEmulatorDevice::isPortForwardingSet()` method is used for checking that the port is forwarded before running UI of our plugin for sensors emulation.
I found the only one way to retrieve this information would be to use the `MerVirtualBoxManager::fetchVirtualMachineInfo()` method.

Is there another way to check port forwarding?

Reply:
Currently only synchronous variant of fetchVirtualMachineInfo exists. You can add an asynchronous version.

Now I also noticed that the way you use freePorts is not valid. fetchVirtualMachineInfo should be fixed to handle ports configured with updatePortForwardingRule separately. The freePorts are ports that the build machinery is free to use, which is not your case - you are setting up a port forwarding rule for your specific use. So you will need to add e.g. otherPorts. And you should also ensure that these two do not conflict.